dipy/tracking/streamlinespeed.pyx

```diff
@@ -301,3 +301,197 @@ def set_number_of_points(streamlines, nb_points=3):
+
+
+cdef double c_dist_to_line(Streamline streamline, np.npy_intp prev,
+                           np.npy_intp next, np.npy_intp curr) nogil:
+    """ Computes the shortest Euclidean distance between a point `curr` and
+        the line passing through `prev` and `next`. """
+
+    cdef:
+        double dn, norm1, norm2
+        np.npy_intp D = streamline.shape[1]
+
+    # Compute cross product of next-prev and curr-next
+    norm1 = c_norm_of_cross_product(streamline[next, 0]-streamline[prev, 0],
+                                    streamline[next, 1]-streamline[prev, 1],
+                                    streamline[next, 2]-streamline[prev, 2],
+                                    streamline[curr, 0]-streamline[next, 0],
+                                    streamline[curr, 1]-streamline[next, 1],
+                                    streamline[curr, 2]-streamline[next, 2])
+
+    # Norm of next-prev
+    norm2 = 0.0
+    for d in range(D):
+        dn = streamline[next, d]-streamline[prev, d]
+        norm2 += dn*dn
+    norm2 = sqrt(norm2)
+
+    return norm1 / norm2
+
+
+cdef np.npy_intp c_compress_streamline(Streamline streamline, Streamline out,
+                                       double tol_error, double max_segment_length) nogil:
+    """ Compresses a streamline (see function `compress_streamlines`)."""
+    cdef:
+        np.npy_intp N = streamline.shape[0]
+        np.npy_intp D = streamline.shape[1]
+        np.npy_intp nb_points = 0
+        np.npy_intp last_added_point_idx = 0
+        np.npy_intp last_success = 0
+        np.npy_intp prev = 0
+        np.npy_intp i, k, next, curr
+        double dn, segment_length
+
+    # Copy first point since it is always kept.
+    for d in range(D):
+        out[0, d] = streamline[0, d]
+
+    last_success = 0
+    nb_points = 1
+
+    for i in range(1, N-1):
+        # Euclidean distance between last added point and current point.
+        segment_length = 0.0
+        for d in range(D):
+            dn = streamline[i, d] - streamline[prev, d]
+            segment_length += dn*dn
+
+        segment_length = sqrt(segment_length)
+        next = i+1
+
+        # Check that each point is not offset by more than `tol_error` mm.
+        for curr in range(prev, i):
+            dist = c_dist_to_line(streamline, prev, next, curr)
+
+            # If the current point's offset is greater than `tol_error`, use the latest success.
+            if dist > tol_error or segment_length > max_segment_length:
+                for d in range(D):
+                    out[nb_points, d] = streamline[last_success, d]
+
+                nb_points += 1
+                prev = i-1
+                last_success = i-1
+                break
+            # The current point's offset is ok, check the next point.
+            else:
+                last_success = i
+
+    # Copy last point since it is always kept.
+    for d in range(D):
+        out[nb_points, d] = streamline[N-1, d]
```

Comment:
Im not sure why, but this line doesn`t seem to work properly.
When I verified if the streamlines had the same end point, the last point of the compressed streamline is in fact the last-1 of the original streamline